### PR TITLE
KAFKA-5355: Test cases to ensure isolation level propagated in delayed fetch

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala [hunk 1]
```diff
@@ -848,8 +853,8 @@ class Log(@volatile var dir: File,
    * @throws OffsetOutOfRangeException If startOffset is beyond the log end offset or before the log start offset
    * @return The fetch data information including fetch starting offset metadata and messages read.
    */
-  def read(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None, minOneMessage: Boolean = false,
-           isolationLevel: IsolationLevel = IsolationLevel.READ_UNCOMMITTED): FetchDataInfo = {
+  def readWithIsolation(startOffset: Long, maxLength: Int, maxOffset: Option[Long] = None, minOneMessage: Boolean = false,
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Shouldn't the public method be simply `read`? We can maybe call the `private[log]` one `readUncommitted`. That will make it very clear when reading the code if it's called somewhere where it should not.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 2]
```diff
@@ -49,13 +49,13 @@ class TransactionsTest extends KafkaServerTestHarness {
...
   override def generateConfigs: Seq[KafkaConfig] = {
-    TestUtils.createBrokerConfigs(numServers, zkConnect, true).map(KafkaConfig.fromProps(_, serverProps()))
+    TestUtils.createBrokerConfigs(numServers, zkConnect).map(KafkaConfig.fromProps(_, serverProps()))
   }
 
   @Before
   override def setUp(): Unit = {
     super.setUp()
-    val numPartitions = 3
+    val numPartitions = 1
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: Why was this changed?

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 3]
```diff
@@ -112,6 +112,53 @@ class TransactionsTest extends KafkaServerTestHarness {
   }
 
   @Test
+  def testReadCommittedConsumerShouldNotSeeUndecidedData(): Unit = {
+    val producer1 = transactionalProducers.head
+    val producer2 = TestUtils.createTransactionalProducer("other", servers)
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: We are leaking this producer. Is there a reason why you cannot use one of the existing `transactionalProducers`?

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 4]
```diff
@@ -112,6 +112,53 @@ class TransactionsTest extends KafkaServerTestHarness {
   }
 
   @Test
+  def testReadCommittedConsumerShouldNotSeeUndecidedData(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala:116]: From a terminology perspective, what is the difference between uncommitted and undecided data?

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 5]
```diff
@@ -378,8 +426,10 @@ class TransactionsTest extends KafkaServerTestHarness {
     val records = new ArrayBuffer[ConsumerRecord[Array[Byte], Array[Byte]]]()
     TestUtils.waitUntilTrue(() => {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: This method seems very similar to `TestUtils.consumeRecords`. Maybe we can consolidate.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala [hunk 6]
```diff
@@ -112,6 +112,53 @@ class TransactionsTest extends KafkaServerTestHarness {
...
+    val records = pollUntilExactlyNumRecords(readCommittedConsumer, 2)
+    records.foreach { record =>
+      assertEquals("x", new String(record.key))
+      assertEquals("1", new String(record.value))
+    }
+
+    // even if we seek to the end, we should not be able to see the undecided data
+    assertEquals(2, readCommittedConsumer.assignment.size)
+    readCommittedConsumer.seekToEnd(readCommittedConsumer.assignment)
+    readCommittedConsumer.assignment.asScala.foreach { tp =>
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/TransactionsTest.scala:157]: Just to double-check, this test failed without the fix?

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 7]
```diff
@@ -1226,11 +1231,10 @@ class LogTest {
     log = new Log(logDir, config, logStartOffset = 0L, recoveryPoint = numMessages + 1, scheduler = time.scheduler,
       brokerTopicStats = brokerTopicStats, time = time)
     val segArray = log.logSegments.toArray
-    for (i <- 0 until segArray.size - 1) {
+    for (i <- segArray.indices.dropRight(1)) {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala]: Instead of `dropRight(1)`, you can call `init` (it's like `tail`, but drops the last element instead of first).

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 8]
```diff
@@ -214,23 +201,32 @@ class ReplicaManagerTest {
...
+      purgatoryName = "Produce", timer, reaperEnabled = false)
+    val mockFetchPurgatory = new DelayedOperationPurgatory[DelayedFetch](
+      purgatoryName = "Fetch", timer, reaperEnabled = false)
+    val mockDeleteRecordsPurgatory = new DelayedOperationPurgatory[DelayedDeleteRecords](
+      purgatoryName = "DeleteRecords", timer, reaperEnabled = false)
+
+    val replicaManager = new ReplicaManager(config, metrics, time, zkUtils, new MockScheduler(time), mockLogMgr,
       new AtomicBoolean(false), QuotaFactory.instantiate(config, metrics, time).follower, new BrokerTopicStats,
-      metadataCache, Option(this.getClass.getName))
+      metadataCache, mockProducePurgatory, mockFetchPurgatory, mockDeleteRecordsPurgatory, Option(this.getClass.getName))
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:215]: Do we want to allow the purgatories to be passed or would it be enough to allow a `Timer` to be passed to `ReplicaManager`? It looks like we are also disabling the reaper, but not sure if that matters. The main advantage is that we would not have to duplicate the constructor.

**Review by ijuma — COMMENTED** [review]
Thanks for the quick PR adding the tests even though the fix was already merged. Looks good overall, just a few minor comments.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 2 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: Having a single partition makes the test assertions simpler. I'll revert this and only use one partition in the new test case.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 4 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala:116]: Admittedly the terminology can be confusing (especially with the overload of "committed" data).  The term "undecided"refers to data which has yet to be aborted or committed. "Uncommitted" in the "read_uncommitted" sense refers to all data in the log, regardless of its transaction commit status (i.e. whether committed, aborted, or undecided). The LSO is intended to keep the read_committed consumer from seeing undecided data.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 3 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala]: It must use a different transactionalId. I will fix the leak.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 6 above)
Comment (hachikuji) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala:157]: Yes.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 8 above)
Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:215]: I had difficulty getting this test working with the reaper enabled. The problem is that it advances the clock asynchronously.

File: core/src/test/scala/integration/kafka/api/TransactionsTest.scala (hunk 4 above)
Comment (ijuma) [reply core/src/test/scala/integration/kafka/api/TransactionsTest.scala:116]: Thanks for the explanation, makes sense.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 8 above)
Comment (ijuma) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:215]: Makes sense.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 9]
```diff
@@ -1330,7 +1330,7 @@ object TestUtils extends Logging {
     waitUntilTrue(() => {
       records ++= consumer.poll(50).asScala
       records.size >= numMessages
-    }, s"Consumed ${records.size} records until timeout instead of the expected $numMessages records", waitTime)
+    }, s"Could not consume $numMessages records before timing out", waitTime)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala]: It seems like it's useful to know how many were consumed in this case, right?

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Yes, but the string is evaluated when `waitUntilTrue` is called, so it always will say you consumed 0 records.

Comment (ijuma) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Good point. We should make the string a by name parameter so that it is evaluated only when used. But it can be done separately if that's easier.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 10]
```diff
@@ -886,6 +889,9 @@ public void handleResponse(AbstractResponse response) {
             Errors error = addOffsetsToTxnResponse.error();
 
             if (error == Errors.NONE) {
+                log.debug("{}Successfully added partition for consumer group {} to transaction", logPrefix,
+                        builder.consumerGroupId());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:893]: Is it worth including information from `offsets`? Also, is this `partition` or `partitions`? Finally, since this is the `AddOffsetsToTxnHandler`, shouldn't we state something about that? Reading this one quickly, it could be confused with the `AddPartitionsToTxnHandler`.

**Review by ijuma — APPROVED** [review]
Thanks for the updates, LGTM. I left a question, but that can be addressed in one of the other PRs, if necessary.
